kv/kvserver: TestReplicaTombstone failed

kv/kvserver.TestReplicaTombstone [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10431502?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10431502?buildTab=artifacts#/) on release-23.1 @ [1a5dc2aad16ca4185afcda75d923804324dcfeb9](https://github.com/cockroachdb/cockroach/commits/1a5dc2aad16ca4185afcda75d923804324dcfeb9):


```
=== RUN   TestReplicaTombstone
--- FAIL: TestReplicaTombstone (40.99s)
=== RUN   TestReplicaTombstone/(4)_(4.1)_raft_messages_to_newer_replicaID_
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaTombstone_(4)_(4.1)_raft_messages_to_newer_replicaID_2468918913
    test_log_scope.go:79: use -show-logs to present logs inline
    client_replica_test.go:3654: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_replica_test.go:3654
        	Error:      	Expect "error sending couldn't accept range_id:64 coordinator_replica:<node_id:1 store_id:1 replica_id:1 type:VOTER_FULL > recipient_replica:<node_id:3 store_id:3 replica_id:4 type:LEARNER > delegated_sender:<node_id:1 store_id:1 replica_id:1 type:VOTER_FULL > priority:REBALANCE type:INITIAL term:9 first_index:41 descriptor_generation:9 queue_on_delegate_len:-1 snap_id:b792b7c5-3e44-4ca3-b924-2e16ba3f5ed2: sender: (n1,s1):1 is not fit to send snapshot for (n1,s1):1; sender term: 10, coordinator term: 9" to match "boom"
        	Test:       	TestReplicaTombstone/(4)_(4.1)_raft_messages_to_newer_replicaID_
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaTombstone_(4)_(4.1)_raft_messages_to_newer_replicaID_2468918913
    --- FAIL: TestReplicaTombstone/(4)_(4.1)_raft_messages_to_newer_replicaID_ (7.44s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaTombstone.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replica_command.go
+++ b/pkg/kv/kvserver/replica_command.go
@@ -2897,7 +2897,7 @@ func (r *Replica) sendSnapshotUsingDelegate(
 
 // validateSnapshotDelegationRequest will validate that this replica can send
 // the snapshot that the coordinator requested. The main reasons a request can't
-// be delegated are if the Generation or Term of the replica is not equal to the
+// be delegated are if the Generation or Term of the replica are less than the
 // Generation or Term of the coordinator's request or the applied index on this
 // replica is behind the truncated index of the coordinator. Note that the request
 // is validated twice, once before "queueing" and once after. This reduces the
@@ -2909,18 +2909,22 @@ func (r *Replica) validateSnapshotDelegationRequest(
 	ctx context.Context, req *kvserverpb.DelegateSendSnapshotRequest,
 ) error {
 	desc := r.Desc()
-	// If the generation has changed, this snapshot may be useless, so don't
-	// attempt to send it.
-	// NB: This is an overly strict check. If other delegates are added to this
-	// snapshot, we don't necessarily need to reject sending the snapshot, however
-	// if there are merges or splits, it is safer to reject.
-	if desc.Generation != req.DescriptorGeneration {
+	// If the delegate doesn't know about a generation change (its index is lower
+	// than the leaseholders) the snapshot it sends may be useless, so don't
+	// attempt to send it and instead return an error.
+	// NB: This is an overly strict check. Some generation changes are acceptable
+	// even if the delegate doesn't know about them. Changes like splits or merges
+	// we don't know about can invalidate the snapshot we would otherwise send. If
+	// the delegates generations is higher than the request generation, then the
+	// snapshot will still be valid since the leaseholder is also on this
+	// generation by now.
+	if desc.Generation < req.DescriptorGeneration {
 		log.VEventf(ctx, 2,
-			"%s: generation has changed since snapshot was generated %s != %s",
+			"%s: generation has changed since snapshot was generated %s < %s",
 			r, req.DescriptorGeneration, desc.Generation,
 		)
 		return errors.Errorf(
-			"%s: generation has changed since snapshot was generated %s != %s",
+			"%s: generation has changed since snapshot was generated %s < %s",
 			r, req.DescriptorGeneration, desc.Generation,
 		)
 	}
@@ -2959,13 +2963,15 @@ func (r *Replica) validateSnapshotDelegationRequest(
 	replTerm := kvpb.RaftTerm(status.Term)
 	r.mu.RUnlock()
 
-	// Delegate has a different term than the coordinator. This typically means
-	// the lease has been transferred, and we should not process this request.
-	// There is a potential race where the leaseholder sends a delegate request
-	// and then the term changes before this request is processed. In that
-	// case this code path will not be checked and the snapshot will still be
-	// sent.
-	if replTerm != req.Term {
+	// Delegate has a lower term than the coordinator. This typically means the
+	// lease has been transferred, and we should not process this request. There
+	// is a potential race where the leaseholder sends a delegate request and then
+	// the term changes before this request is processed. In that case the
+	// snapshot will still be sent. There isn't a problem sending a snapshot with
+	// a lower term, however the new leaseholder may think we need a snapshot and
+	// send an additional one. This check attempts to minimize the change of the
+	// double snapshot being sent.
+	if replTerm < req.Term {
 		log.Infof(
 			ctx,
 			"sender: %v is not fit to send snapshot for %v; sender term: %v coordinator term: %v",

